Add ability to specify transaction config in Neo4jGraphQLContext

luffy1727 opened the pull request:
# Description

Adding an option to specify transaction properties to Neo4jGraphQLContext

## Complexity

Complexity: Low

# Checklist

The following requirements should have been met (depending on the changes in the branch):

- [ ] Documentation has been updated
- [ ] TCK tests have been updated
- [ ] Integration tests have been updated
- [ ] Example applications have been updated
- [ ] New files have copyright header
- [ ] CLA (https://neo4j.com/developer/cla/) has been signed

luffy1727 committed:
adding an option to specify transaction timeout

angrykoala commented on packages/graphql/src/classes/Executor.ts:
```diff
@@ -280,4 +277,4 @@ export class Executor {
         return transaction.run(queryToRun, parameters);
     }
-}
+}
```
Any reason for the file ending change?

Obviously not important, but If not consistent across the PRs, this cause unnecessary noise in the diff

Could this be caused by an issue in the formatting tool?

luffy1727 replied:
This was my mechanical error. Nothing related to the formatting tool. My bad

angrykoala commented on packages/graphql/src/types/neo4j-graphql-context.ts:
```diff
@@ -51,6 +51,15 @@ export interface Neo4jGraphQLContext extends Neo4jGraphQLContextInterface {
      * This can be used to output information to the query log not related to the query itself.
      * Will be ignored if {@link executionContext} is an instance of a transaction.
+     * 
+     * @deprecated This method will be removed in version ???.0. Please, use {@link transaction} instead.
```
If we are not sure of the version, maybe ignore the version from the message and just mention it will be removed?

Also, maybe detail the exact field? `transaction.metadata`?

angrykoala commented on packages/graphql/src/classes/Executor.ts:
```diff
@@ -205,9 +200,11 @@ export class Executor {
                 ...this.transactionMetadata,
                 type: "user-transpiled",
             },
+            timeout: this.transaction.timeout ?? 60 * 1000,
```
Do we want to add a default to timeout?

The current behaviour is to have this undefined, no?

angrykoala commented on packages/graphql/src/classes/Executor.ts:
```diff
@@ -95,20 +88,23 @@ export class Executor {
 
     private cypherParams: Record<string, unknown>;
-    private transactionMetadata: Record<string, unknown>;
+    private transactionMetadata: Record<string, unknown>; // deprecate this
```
```suggestion
    private transactionMetadata: Record<string, unknown>; // TODO: deprecate this
```

luffy1727 committed:
adding tests for user defined transaction properties

mjfwebb commented on packages/graphql/src/types/neo4j-graphql-context.ts:
```diff
@@ -51,6 +51,15 @@ export interface Neo4jGraphQLContext extends Neo4jGraphQLContextInterface {
      * This can be used to output information to the query log not related to the query itself.
      * Will be ignored if {@link executionContext} is an instance of a transaction.
+     * 
+     * @deprecated This method will be removed in later version. Please, use {@link transaction} instead.
```
```suggestion
     * @deprecated This method will be removed in a future version. Please, use {@link transaction} instead.
```

luffy1727 committed:
revert changes for TransactionConfig type

- reverting changes for TransactionConfig type by making it private
- removing transactionMetadata from Executor

mjfwebb commented on packages/graphql/src/schema/resolvers/composition/wrap-query-and-mutation.ts:
```diff
@@ -104,7 +104,10 @@ export const wrapQueryAndMutation =
             sessionConfig: context.sessionConfig,
             cypherParams: context.cypherParams,
-            transactionMetadata: context.transactionMetadata,
+            transaction: {
+                timeout: context.transaction?.timeout,
+                metadata: {...context.transaction?.metadata, ...context.transactionMetadata},
+            },
```
```suggestion
            transaction: {
                ...context.transaction
                // TODO: Remove explicit metadata property assignment when transactionMetadata is removed
                metadata: {...context.transaction?.metadata, ...context.transactionMetadata},
            },
```

I think this will allow us to accept anything, even if the `UserTransactionConfig` will need to be maintained. From the perspective of JS, or asserted TS, we would just allow all additional transaction data that may be added in the future to the driver.

luffy1727 replied:
You are right. Changed. 🫡

luffy1727 committed:
add changeset

mjfwebb commented on .changeset/slow-plants-beam.md:
```diff
@@ -0,0 +1,5 @@
+"@neo4j/graphql": minor
+---
+
+Add an ability to specify transaction configuration. e.g. timeout
```
Can you add to this changeset a little code example?

mjfwebb commented on packages/graphql/src/classes/Executor.ts:
```diff
@@ -82,9 +84,14 @@ export type ExecutorConstructorParam = {
     sessionConfig?: SessionConfig;
     cypherParams?: Record<string, unknown>;
-    transactionMetadata?: Record<string, unknown>;
+    transaction?: UserTransactionConfig;
```
transactionMetadata should still be here, right? Just deprecated

luffy1727 replied:
it is merged in transaction.metadata in `wrap-query-and-mutation` so that this file only deals with the new transaction config.

mjfwebb replied:
Understood!

mjfwebb reviewed (dismissed):
Good stuff 👍

angrykoala commented on packages/graphql/src/utils/execute.test.ts:
```diff
@@ -271,4 +274,193 @@ describe("execute", () => {
+                    }),
+                }).instance(),
+            });
+            // console.log(executeResult)
```
```suggestion
```

angrykoala commented on .changeset/slow-plants-beam.md:
```diff
@@ -0,0 +1,24 @@
+"@neo4j/graphql": minor
+---
+
+Add an ability to specify transaction configuration. e.g. timeout
```
```suggestion
Add the ability to specify transaction configuration. e.g. timeout
```

angrykoala reviewed (dismissed):
Two very small changes and it is good to merge

luffy1727 committed:
Merge branch 'dev' into adding-transaction-timeout-option
